plugins/inputs/intel_powerstat/services/rapl.go

```diff
@@ -0,0 +1,274 @@
+	}
+	defer maxPowerFile.Close()
+	maxPower, _, err := r.fs.ReadFileToFloat64(maxPowerFile)
+
+	return maxPower, err
+}
+
+// GetDramMaxEnergyRangeUj retrieves dram energy range value from max_energy_range_uj file.
+func (r *RaplServiceImpl) GetDramMaxEnergyRangeUj(socketID string) (float64, error) {
+	dramPath := fmt.Sprintf(IntelRaplDramPartialPath, IntelRaplPath, socketID, r.dramFolders[socketID])
+	dramEnergyPath := fmt.Sprintf(MaxRangeUjPartialPath, dramPath)
+	maxDramEnergyFile, err := os.Open(dramEnergyPath)
+	if err != nil {
+		return 0, fmt.Errorf("error opening max_energy_range_uj file on path %s, err: %v", dramEnergyPath, err)
+	}
+	defer maxDramEnergyFile.Close()
+	maxDramEnergy, _, err := r.fs.ReadFileToFloat64(maxDramEnergyFile)
+
+	return maxDramEnergy, err
+}
+
+// GetMaxEnergyRangeUj retrieves energy range value from max_energy_range_uj file.
+func (r *RaplServiceImpl) GetMaxEnergyRangeUj(socketID string) (float64, error) {
+	raplPath := fmt.Sprintf(IntelRaplSocketPartialPath, IntelRaplPath, socketID)
+	maxEnergyPath := fmt.Sprintf(MaxRangeUjPartialPath, raplPath)
+	maxEnergyFile, err := os.Open(maxEnergyPath)
+	if err != nil {
+		return 0, fmt.Errorf("error opening max_energy_range_uj file on path %s, err: %v", maxEnergyPath, err)
+	}
+	defer maxEnergyFile.Close()
+	maxEnergy, _, err := r.fs.ReadFileToFloat64(maxEnergyFile)
+
+	return maxEnergy, err
+}
+
+// getDramFolders returns Dram folder names from RAPL.
+func (r *RaplServiceImpl) getDramFolders() map[string]string {
+	return r.dramFolders
+}
+
+func (r *RaplServiceImpl) prepareData() {
+	intelRaplPrefix := "intel-rapl:"
+	intelRapl := fmt.Sprintf("%s%s", intelRaplPrefix, "[0-9]*")
+	raplPaths, err := r.fs.GetStringsMatchingPatternOnPath(fmt.Sprintf("%s/%s", IntelRaplPath, intelRapl))
+	if err != nil {
+		if r.isFirstError {
+			r.isFirstError = false
+			r.log.Errorf("error while preparing RAPL data: %v", err)
+		} else {
+			r.log.Debugf("error while preparing RAPL data: %v", err)
+		}
+
+		r.data = make(map[string]*data.RaplData)
+		return
+	}
+
+	r.isFirstError = true
+	// If RAPL exists initialize data map (if it wasn't initialized before).
+	if len(r.data) == 0 {
+		for _, raplPath := range raplPaths {
+			socketID := strings.TrimPrefix(filepath.Base(raplPath), intelRaplPrefix)
+			r.data[socketID] = &data.RaplData{
+				SocketCurrentEnergy: 0,
+				DramCurrentEnergy:   0,
+				SocketEnergy:        0,
+				DramEnergy:          0,
+				ReadDate:            0,
+			}
+		}
+	}
+}
+
+func (r *RaplServiceImpl) findDramFolders() {
+	intelRaplPrefix := "intel-rapl:"
+	intelRaplDram := fmt.Sprintf("%s%s", intelRaplPrefix, "[0-9]*[0-9]*")
+	// Clean existing map
+	r.dramFolders = make(map[string]string)
+
+	for socketID := range r.data {
+		path := fmt.Sprintf(IntelRaplSocketPartialPath, IntelRaplPath, socketID)
+		pathsToRaplFolders, err := r.fs.GetStringsMatchingPatternOnPath(fmt.Sprintf("%s/%s", path, intelRaplDram))
```

Comment:
Again this should not fail, if it does, there is a big problem anyway.

Reply:
I understand, this is almost impossible to happen that this pattern is not well-formatted but for security reasons and for the sake of handling every error we would like to leave it as it is.